examples/echo/main.c

```diff
@@ -239,6 +274,24 @@ static void handle_rx (YggWorker   *worker,
   g_bytes_unref (data);
 }
 
+/**
+ * A #YggCancelFunc callback that is invoked when the worker receives a cancel
+ * message from the dispatcher.
+ */
+static void handle_cancel (YggWorker *worker,
+                           gchar     *addr,
+                           gchar     *id,
+                           gchar     *cancel_id,
+                           gpointer   user_data)
+{
+  g_debug("handle_cancel");
+  GCancellable *cancellable = g_hash_table_lookup (active_sources, cancel_id);
+  if (cancellable != NULL) {
+    g_cancellable_cancel (cancellable);
+    g_hash_table_remove (active_sources, cancel_id);
+  }
+}
```

Comment:
This handle has never been called, when I tried to test it. I can see following message on `yggd` terminal:

```
[yggd] 2024/01/04 13:31:45 /home/jhnidek/github/redhat-insights/yggdrasil/cmd/yggd/client.go:310: received message e012f32e-5b42-48f5-8fdd-1e9d31a5e0a4
[yggd] 2024/01/04 13:31:45 /home/jhnidek/github/redhat-insights/yggdrasil/cmd/yggd/client.go:311: command: cancel
[yggd] 2024/01/04 13:31:45 /home/jhnidek/github/redhat-insights/yggdrasil/cmd/yggd/client.go:312: Control message: &{command e012f32e-5b42-48f5-8fdd-1e9d31a5e0a4  1 2024-01-04 13:31:45.131280883 +0100 CET [123 34 99 111 109 109 97 110 100 34 58 34 99 97 110 99 101 108 34 44 34 97 114 103 117 109 101 110 116 115 34 58 123 34 100 105 114 101 99 116 105 118 101 34 58 34 101 99 104 111 34 44 34 109 101 115 115 97 103 101 73 68 34 58 34 48 53 102 57 102 97 52 55 45 51 50 99 48 45 52 97 53 49 45 98 50 102 102 45 57 49 101 52 53 51 52 53 52 97 56 100 34 125 125]}
[yggd] 2024/01/04 13:31:45 /home/jhnidek/github/redhat-insights/yggdrasil/cmd/yggd/client.go:349: cancelling message...
[yggd] 2024/01/04 13:31:45 /home/jhnidek/github/redhat-insights/yggdrasil/internal/transport/mqtt.go:85: cannot receive control message: cannot process control message: cannot dispatch cancel message: cannot call Cancel method with message 05f9fa47-32c0-4a51-b2ff-91e453454a8d on worker echo: No such method “Cancel”
```

Reply:
BTW: Canceling works as expected for echo worker written in Go using same `yggd`.

Reply:
I suspect this might be because of the version of `com.redhat.Yggdrasil1.Worker1.xml` that the client is finding on your system. Do you have yggdrasil installed from Fedora repositories? I don't think that version has the Cancel method defined, and I wonder if the broker doesn't attempt to invoke the method on the client if the introspection file does not define it. Maybe try pointing your custom build of libygg at a `main` branch of yggdrasil using the `worker_introspection_file` meson build option?

Reply:
Actually, now that I think about it, it's probably the version of `libygg.so.0` that's getting loaded. The version that is shipping in Fedora does not include the message cancellation functionality, so it wouldn't set the handle_cancel callback. If you make sure to put the locally compiled `libygg.so.0` into the `LD_LIBRARY_PATH`, that version includes cancellation support and will attempt to set the handle_cancel callback.

Reply:
Yeah, the `/usr/share/dbus-1/interfaces/com.redhat.Yggdrasil1.Worker1.xml` installed from Fedora rpm does not contain `Cancel` method. Re-configuring meson using: `meson configure builddir -Dworker_interface_file=/path/to/yggdrasil/ipc/com.redhat.Yggdrasil1.Worker1.xml` and rebuilding `libygg` using: `meson compile -C builddir` fixed this issue.